url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/quantum/SystemMatrix.java

```diff
@@ -0,0 +1,233 @@
+/*******************************************************************************
+ * Copyright (c) 2013 Martin Marinov.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the GNU Public License v3.0
+ * which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/gpl.html
+ * 
+ * Contributors:
+ *     Martin - initial API and implementation
+ ******************************************************************************/
+package martin.quantum;
+
+import java.util.HashMap;
+
+import martin.math.Complex;
+import martin.math.MathExpression;
+import martin.math.MathNumber;
+import martin.math.MathsItem;
+import martin.operators.AddCoeffTogether;
+import martin.operators.Operator;
+import martin.quantum.tools.Tools;
+
+public class SystemMatrix {
+
+	public final int mNumbQubits;
+	public MathsItem[] coeff;
+	public final boolean[] measured;
+	public final int size;
+
+	/**
+	 * Initialises a qubit system with predetermined coefficients.
+	 * @param items the coefficients that the {@link SystemMatrix} will have. The id is the binary
+	 * representation of the braket. The number of items should be 2^n where n is the number of qubits in
+	 * the system.
+	 * @throws Exception
+	 */
+	public SystemMatrix(final MathsItem[] items) throws Exception {
+		this(Tools.powerOfTwo(items.length));
+		for (int i = 0; i < items.length; i++)
+			coeff[i].multiply(items[i].clone());
+		simplify();
+	}
+	
+	/**
+	 * Initialises a system matrix with a certain number of qubits. All coefficients are set to 1.
+	 * @param qubitCount
+	 */
+	public SystemMatrix(int qubitCount) {
+		mNumbQubits = qubitCount;
+		size = 1 << qubitCount;
+		coeff = new MathsItem[size];
+		measured = new boolean[qubitCount];
+		for (int i = 0; i < qubitCount; i++) measured[i] = false;
+		for (int i = 0; i < size; i++) {
+			coeff[i] = new MathExpression();
+			coeff[i].add(new MathNumber(1));
+		}
+	}
+	
+	@Override
+	protected Object clone() throws CloneNotSupportedException {
+		final SystemMatrix clone = new SystemMatrix(mNumbQubits);
+		
+		for (int i = 0; i < size; i++)
+			clone.coeff[i] = coeff[i].clone();
+		
+		for (int i = 0; i < mNumbQubits; i++)
+				clone.measured[i] = measured[i];
+	
+		return clone;
+	}
+
+	public void setCoeff(final MathsItem c, int... indeces) throws Exception {
+		if (indeces.length != mNumbQubits)
+			throw new Exception("Not enough indeces supplied");
+
+		int id = 0;
+		for (int i = 0; i < mNumbQubits; i++)
+			if (indeces[i] != 0)
+				id += indeces[i] * (1 << i);
+		coeff[id] = c;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;c&#39; (index: 0 | type: MathsItem). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;indeces&#39;. The same parameter type is validated here for example: url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/math/MathExp.java:23. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).